File: my-code/0045-jump-game-ii/solution.cpp
class Solution {
public:
    int jump(vector<int>& nums) {

        //tabulation:
        int n = nums.size();
        vector<int> dp(nums.size(),0);
        for(int i=n-2;i>=0;i--){
               int mini = INT_MAX;
               int step = nums[i];
               if(step>0){
                   for(int j=1;j<=step;j++){
                       if(i+j<nums.size()){
                           mini = min(mini,dp[i+j]);
                       }
                   }
               }
               if(mini==INT_MAX){
                    dp[i]=mini;
               }
               else{
                   dp[i]=mini+1;
               }

        }
        return dp[0];
    }
};

File: my-code/0045-jump-game-ii/solution.java
class Solution {
    public int jump(int[] arr) {
        int dp[] = new int[arr.length];

        // dp[arr.length-1] = ;

        for(int idx = arr.length-2 ; idx >=0 ; idx--){
            int steps = arr[idx];
        
            int min = Integer.MAX_VALUE;
            if(steps > 0){
                for(int i = 1 ; i <= steps ;i++){
                    if(idx + i < arr.length)
                        min = Math.min(min,dp[idx+i]);
                }
            }
            dp[idx] = min == Integer.MAX_VALUE ? min : min+1;
          
        }
        return dp[0];
    }
}
